lib/crc32_arm.h

```diff
@@ -0,0 +1,162 @@
+    return veorq_u64(tmp_l, veorq_u64(to, tmp_h));
+}
+
+static inline uint64x2_t fold_64b(uint64x2_t from, const uint64x2_t constant)
+{
+    uint64x2_t tmp = from;
+
+    /* from[0]*constant[1], in reflected domain ,from[0] is the high 64bit */
+    from = (uint64x2_t)vmull_p64((poly64_t)vgetq_lane_u64(from, 0), (poly64_t)vgetq_lane_u64(constant, 1));
+    from = veorq_u64(__shift_p128_right(tmp, 8), from);
+
+    /* Fold 96 => 64 bits */
+    tmp = __shift_p128_right(from, 4);
+    from = (uint64x2_t)vmull_p64((poly64_t)vgetq_lane_u64(vandq_u64(from, mask32), 0), (poly64_t)vgetq_lane_u64(constant, 0));
+    return veorq_u64(tmp, from);
+}
+
+static inline uint64_t barrett_reduction(uint64x2_t data, const uint64x2_t p_q)
+{
+    uint64x2_t tmp;
+
+    tmp = data;
+    data = (uint64x2_t)vmull_p64((poly64_t)vgetq_lane_u64(vandq_u64(data, mask32), 0), (poly64_t)vgetq_lane_u64(p_q, 1));
+    data = (uint64x2_t)vmull_p64((poly64_t)vgetq_lane_u64(vandq_u64(data, mask32), 0), (poly64_t)vgetq_lane_u64(p_q, 0));
+
+    return vgetq_lane_u32((uint32x4_t)veorq_u64(tmp, data), 1);
+}
+
+static u32 ATTRIBUTES
+FUNCNAME_ALIGNED(u32 remainder, const unsigned char *p, size_t length)
+{
+    uint64x2_t *p_data = (uint64x2_t *)p;
+    uint64_t remain_len = length, crc = remainder;
+
+    uint64x2_t x0, x1, x2, x3;
+    uint64x2_t y0, y1, y2, y3;
+
+    /* expand crc to 128bit */
+    y0 = vcombine_u64((uint64x1_t)crc, (uint64x1_t)0ULL);
+
+    /* load first 64B */
+    x0 = *p_data++; x1 = *p_data++;
+    x2 = *p_data++; x3 = *p_data++;
+    remain_len -= 64;
+
+    x0 = x0 ^ y0; /* x0 ^ crc */
+
+    /* 1024bit --> 512bit loop */
+    while(remain_len >= 64) {
+        /* load 64 bytes */
+        y0 = *p_data++; y1 = *p_data++;
+        y2 = *p_data++; y3 = *p_data++;
+
+        /* fold 1024bit --> 512bit */
+        x0 = fold_128b(y0, x0, foldConstants_p4);
+        x1 = fold_128b(y1, x1, foldConstants_p4);
+        x2 = fold_128b(y2, x2, foldConstants_p4);
+        x3 = fold_128b(y3, x3, foldConstants_p4);
+
+        remain_len -= 64;
+    }
+
+    /* folding 512bit --> 128bit */
+    x1 = fold_128b(x1, x0, foldConstants_p1);
+    x2 = fold_128b(x2, x1, foldConstants_p1);
+    x0 = fold_128b(x3, x2, foldConstants_p1);
+
+    /*
+     * now we have folded the whole data buffer to 128bits
+     * next target is to fold 128bits -> 64bits.
+     * this is done in two steps:
+     * 1. (128+32)bits -> 96bits (32bits(x^32) are appended)
+     * 2. 96bits -> 64bits
+     */
+    x0 = fold_64b(x0, foldConstants_p0);
+
+    /* calc crc using barrett reduction method */
+    remainder = barrett_reduction(x0, foldConstants_br);
+
+    return remainder;
+}
+
+/*
+ *
+ * Note: on unaligned ends of the buffer, we fall back to crc32_slice1() instead
+ * of crc32_slice8() because only a few bytes need to be processed, so a smaller
+ * table is preferable.
+ */
+static u32 ATTRIBUTES
+FUNCNAME(u32 remainder, const u8 *buffer, size_t nbytes)
```

Comment:
No need for the FUNCNAME abstraction if this header isn't going to be include multiple times; it should just be called `crc32_pmull`.  Similarly for FUNCNAME_ALIGNED and ATTRIBUTES.